### PR TITLE
Add dbname to DiskQuotaWorkerEntry

### PR DESCRIPTION
**Pull request by zhrt123**
## Problem
Previously, we passed dbname by bgw_name to bgworker. But refer to https://github.com/greenplum-db/gpdb/blob/f182228992b62e2023e2fac5b4971406abd35c9d/src/backend/postmaster/bgworker.c#L385-L386, bgw_name is copied by ascii_safe_strlcpy(), which replaces non-ASCII bytes with '?'. If the database name is non-ASCII, the bgworker can not connect to the correct database and raises an error. #323 

## Solution
Pass dbname by bgw_main_arg.

### CODE REVIEW

File: src/diskquota.c [hunk 1]
```diff
@@ -1381,7 +1384,11 @@ start_worker(DiskquotaDBEntry *dbEntry)
 		result = INVALID_DB;
 		goto Failed;
 	}
-	snprintf(worker.bgw_name, sizeof(worker.bgw_name), "%s", dbname);
+	LWLockAcquire(diskquota_locks.workerlist_lock, LW_EXCLUSIVE);
+	namestrcpy(&(dq_worker->dbname), dbname);
```
Comment (fanfuxiaoran) [inline src/diskquota.c:1388]: LGTM.  However,  the dbname is not needed to start the bgworker, so we should not get it here and pass it to diskquota worker through shared memory. Instead, the diskquota worker can get it through `get_db_name` by itself.

File: src/diskquota.c [hunk 2]
```diff
@@ -1381,7 +1384,11 @@ start_worker(DiskquotaDBEntry *dbEntry)
 		result = INVALID_DB;
 		goto Failed;
 	}
-	snprintf(worker.bgw_name, sizeof(worker.bgw_name), "%s", dbname);
+	LWLockAcquire(diskquota_locks.workerlist_lock, LW_EXCLUSIVE);
```
Comment (beeender) [inline src/diskquota.c:1387]: Why lock is needed here but not at line 1361?

File: src/diskquota.c (hunk 1 above)
Comment (beeender) [reply src/diskquota.c:1388]: This sounds better, can we do that?

**Comment by beeender** [comment]
BTW, since this is related to #323 , it is better we have the issue number in the PR/commit message.

File: src/diskquota.c (hunk 1 above)
Comment (zhrt123) [reply src/diskquota.c:1388]: Can the bgworker run get_db_name() before BackgroundWorkerInitializeConnection()?

File: src/diskquota.c (hunk 2 above)
Comment (zhrt123) [reply src/diskquota.c:1387]: Maybe we can remove it.

**Comment by zhrt123** [comment]
> This sounds better, can we do that?

We should give bgworker a name.

File: src/diskquota.c (hunk 1 above)
Comment (fanfuxiaoran) [reply src/diskquota.c:1388]: >Can the bgworker run get_db_name() before BackgroundWorkerInitializeConnection()?

Comment (fanfuxiaoran) [reply src/diskquota.c:1388]: In gpdb7,  a filed ` char            bgw_extra[BGW_EXTRALEN];` is added, we can use it to pass dbname.

File: src/diskquota.c [hunk 3]
```diff
@@ -321,7 +321,12 @@ define_guc_variables(void)
 void
 disk_quota_worker_main(Datum main_arg)
 {
-	char *dbname = MyBgworkerEntry->bgw_name;
+	char dbname[NAMEDATALEN];
+
+	MyWorkerInfo = (DiskQuotaWorkerEntry *)DatumGetPointer(MyBgworkerEntry->bgw_main_arg);
+	Assert(MyWorkerInfo != NULL);
+
+	memcpy(dbname, MyWorkerInfo->dbname.data, NAMEDATALEN);
```
Comment (fanfuxiaoran) [inline src/diskquota.c:329]: Can use the `MyWorkerInfo->dbname.data` directly instead of copying it? And can use `NameStr` to get the dbname.data.

**Review by fanfuxiaoran — APPROVED** [review]

File: src/diskquota.c (hunk 1 above)
Comment (fanfuxiaoran) [reply src/diskquota.c:1388]: after discussion, we decide to save the dbname in shared memory as the pr did to keep it has the same behaviors between gpdb6 and gpdb7.

**Review by fanfuxiaoran — APPROVED** [review]

File: tests/regress/expected/test_bgworker_name.out
[the remark this answers is not in the document]
Comment (zhrt123) [reply tests/regress/expected/test_bgworker_name.out:1]: Renamed

File: src/diskquota.c (hunk 3 above)
Comment (beeender) [reply src/diskquota.c:329]: one change

```
const char* dbname = MyWorkerInfo->dbname.data;
```

Comment (zhrt123) [reply src/diskquota.c:329]: I think it is better to use local memory to store dbname.

**Review by yihong0618 — APPROVED** [review]

**Review by beeender — APPROVED** [review]

**Comment by zhrt123** [comment]
Previously, we passed dbname by bgw_name to bgworker. But refer to https://github.com/greenplum-db/gpdb/blob/f182228992b62e2023e2fac5b4971406abd35c9d/src/backend/postmaster/bgworker.c#L385-L386, bgw_name is copied by ascii_safe_strlcpy(), which replaces non-ASCII bytes with '?'. If the database name is non-ASCII, the bgworker can not connect to the correct database and raises an error. issue: https://github.com/greenplum-db/diskquota/issues/323 . To solve this problem, we should pass dbname by bgw_main_arg to the main function of the bgworker.

**Pull request merged**